### PR TITLE
Stuff

### LINKED ISSUE #666: support `type_check_only` (`function` is not exported from `builtins` at runtime)
```py
function  # no error
```

### LINKED ISSUE #690: `functools.wraps` should return an intersection, not a `_Wraps`
something like:
```py
class _Wrapped(Protocol[_PWrapped, _RWrapped]):
    __wrapped__: Callable[_PWrapped, _RWrapped]

C = TypeVar("C", bound=Callable)
class _Wrapper(Generic[_PWrapped, _RWrapped]):
    def __call__(self, f: C) -> _Wrapped[_PWrapped, _RWrapped] & C: ...
```

well, `_Wrapper` is actually a `partial`, soooooo

### LINKED ISSUE #739: bare literals/tuple literals don't work with new type alias syntax
```py
type Digit = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 # error: valid-type
type X = (int, str) # error: valid-type
```

### LINKED ISSUE #765: disable promotions by default

### LINKED ISSUE #766: fix the default values in the docs

### LINKED ISSUE #769: crash with cached intersectiontype of type alias
totally needs;
```py
        self.items = flatten_nested_unions(items, handle_type_alias_type=False, type_type=IntersectionType)
```

### LINKED ISSUE #773: crash with typealias in regex usage
```py
import re

Regex = re.Pattern[str]


pattern: Regex
re.search(pattern, "")
```

### LINKED ISSUE #776: remove `ellipsis` from `builtins`

### PR DESCRIPTION
**Pull request by KotlinIsland**
image

- typed `functools.wraps` correctly resolves #690
- typed `builtins.ellipsis`/`types.EllipsisType` correctly resolves #776
- disable `bytearray` and `memoryview` promotions by default  resolves #765
- support `typing.type_check_only` resolves #666
- support default values on type parameters
- support based type notation in type alias statements resolves #739
- fixed crash with `re` plugin resolves #773
- fixed crash with `Intersection` cache resolves #769
- fixed default configuration in the docs resolves #766
- upgrade messages on certain types #476

### CODE REVIEW

File: mypy/typeshed/stdlib/types.pyi [hunk 1]
```diff
@@ -602,11 +602,11 @@ if sys.version_info >= (3, 9):
         def __getattr__(self, name: str) -> Any: ...
 
 if sys.version_info >= (3, 10):
-    # This annotation allows `NoneType` to be used as a value of `type[None]`
-    NoneType: Final[type[None]] = ...
+    # ideally, this would have a `: TypeAlias`, but it makes mypy output
+    #  an invalid error on this line
+    NoneType = type(None)
```
Comment (DetachHead) [inline mypy/typeshed/stdlib/types.pyi:607]: can it use the new type alias syntax instead

File: test-data/unit/check-based-intersection.test [hunk 2]
```diff
@@ -341,3 +341,32 @@ reveal_type(x)  # N: Revealed type is "__main__.A & __main__.B & __main__.C | __
...
+
+
+[case testTypeApplication]
+import types
+from typing import Generic, TypeVar, List
+T = TypeVar("T")
+List[int & str]  # E: `X & Y` syntax for intersections requires "__future__.annotations" or quoted types  [valid-type]
+List[int | str]  # E: `X | Y` syntax for unions requires Python 3.10  [valid-type]
+a: List[int & str]  # E: `X & Y` syntax for intersections requires "__future__.annotations" or quoted types  [valid-type]
+b: List[int | str]  # E: `X | Y` syntax for unions requires Python 3.10  [valid-type]
```
Comment (DetachHead) [inline test-data/unit/check-based-intersection.test:353]:
```suggestion
List[int | str]  # E: `X | Y` syntax for unions requires Python 3.10 or above  [valid-type]
a: List[int & str]  # E: `X & Y` syntax for intersections requires "__future__.annotations" or quoted types  [valid-type]
b: List[int | str]  # E: `X | Y` syntax for unions requires Python 3.10 or above  [valid-type]
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
 
 ## [Unreleased]
+### Added
+- typed `functools.wraps` correctly
+- typed `builtins.ellipsis`/`types.EllipsisType` correctly
+- disable `bytearray` and `memoryview` promotions by default
+- support `typing.type_check_only`
+- support default values on type parameters
+- support based type notation in type alias statements
+- support `basedtyping.FunctionType`
```
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply CHANGELOG.md:11]: `as_functiontype` is always needed when ever something is a `Callable`

File: mypy/typeanal.py [hunk 4]
```diff
@@ -661,7 +663,7 @@ def try_analyze_special_unbound_type(self, t: UnboundType, fullname: str) -> Typ
         elif fullname == "typing._NamedCallable":
             return self.analyze_callable_type(t, fullname)
         elif fullname in {"types.FunctionType", "types.BuiltinFunctionType"}:
-            if not self.always_allow_new_syntax:
+            if not self.always_allow_new_syntax and not self.python_3_12_type_alias:
```
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply mypy/typeanal.py:666]: trueee

File: mypy/typeshed/stdlib/types.pyi (hunk 1 above)
Comment (KotlinIsland) [reply mypy/typeshed/stdlib/types.pyi:607]: no, the real isn't a type alias

**Review by DetachHead — APPROVED** [review]

**Pull request merged**